apollo-router/src/ageing_priority_queue.rs

```diff
@@ -49,28 +52,29 @@ impl<T> AgeingPriorityQueue<T>
 where
     T: Send + 'static,
 {
-    pub(crate) fn soft_bounded(soft_capacity: usize) -> Self {
+    pub(crate) fn bounded(capacity: usize) -> Self {
         Self {
             // Using unbounded channels: callers must use `is_full` to implement backpressure
             inner_queues: std::array::from_fn(|_| crossbeam_channel::unbounded()),
             queued_count: AtomicUsize::new(0),
-            soft_capacity,
+            capacity,
         }
     }
 
     pub(crate) fn queued_count(&self) -> usize {
         self.queued_count.load(Ordering::Relaxed)
     }
 
-    pub(crate) fn is_full(&self) -> bool {
-        self.queued_count() >= self.soft_capacity
-    }
-
     /// Panics if `priority` is not in `AVAILABLE_PRIORITIES`
-    pub(crate) fn send(&self, priority: Priority, message: T) {
-        self.queued_count.fetch_add(1, Ordering::Relaxed);
+    pub(crate) fn send(&self, priority: Priority, message: T) -> Result<(), QueueIsFullError> {
+        self.queued_count
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |previous_count| {
+                (previous_count < self.capacity).then_some(previous_count + 1)
+            })
+            .map_err(|_| QueueIsFullError)?;
         let (inner_sender, _) = &self.inner_queues[index_from_priority(priority)];
-        inner_sender.send(message).expect("disconnected channel")
+        inner_sender.send(message).expect("disconnected channel");
+        Ok(())
```

Comment:
I know this isn't new `expect`, but now that return an error from the `send` function, could we not return an error here rather than `expect` the `send` to work?

Reply:
I’ve made this change, but the caller in `compute_jobs.rs` now calls `expect` instead. There’s a comment explaining why this can never panic.